File: python/test/state_representation/space/test_jacobian.py
import unittest
import copy

import numpy as np
from state_representation import Jacobian, CartesianPose, CartesianTwist, CartesianWrench, JointVelocities

JACOBIAN_METHOD_EXPECTS = [
    'Random',
    'rows',
    'row',
    'cols',
    'col',
    'data',
    'get_joint_names',
    'get_frame',
    'get_reference_frame',
    'set_joint_names',
    'set_reference_frame',
    'set_data',
    'is_compatible',
    'initialize',
    'transpose',
    'inverse',
    'pseudoinverse',
    'solve',
    'copy',
]


class TestJacobian(unittest.TestCase):
    def assert_np_array_equal(self, a, b):
        self.assertListEqual(list(a), list(b))

    def assert_np_array_almost_equal(self, a, b):
        a = list(a)
        b = list(b)
        for i in range(len(a)):
            self.assertAlmostEqual(a[i], b[i])

    def test_callable_methods(self):
        methods = [m for m in dir(Jacobian) if callable(getattr(Jacobian, m))]
        for expected in JACOBIAN_METHOD_EXPECTS:
            self.assertIn(expected, methods)

    def test_constructors(self):
        A = Jacobian("robot", 7, "ee")
        Jacobian(A)
        Jacobian("robot", ["joint_0", "joint_1"], "ee", "robot")
        Jacobian.Random("robot", ["joint_0", "joint_1"], "ee", "robot")
        B = Jacobian.Random("robot", 7, "ee")
        B.copy()

    def test_copy(self):
        state = Jacobian.Random("robot", 7, "ee")
        for state_copy in [copy.copy(state), copy.deepcopy(state)]:
            self.assertEqual(state.get_frame(), state_copy.get_frame())
            self.assertEqual(state.get_reference_frame(), state_copy.get_reference_frame())
            self.assertListEqual(state.get_joint_names(), state_copy.get_joint_names())
            self.assert_np_array_equal(state.data().flatten(), state_copy.data().flatten())

    def test_getters(self):
        jac = Jacobian("jacobian", 3, "ee", "robot")
        self.assertEqual(jac.cols(), 3)
        jac.set_joint_names(["joint__0", "joint__1", "joint__2"])
        self.assertListEqual(jac.get_joint_names(), ["joint__0", "joint__1", "joint__2"])
        self.assertEqual(jac.get_frame(), "ee")
        self.assertEqual(jac.get_reference_frame(), "robot")

        data = np.random.rand(6, 3)
        jac.set_data(data)
        self.assert_np_array_equal(jac.col(1), data[:, 1])
        self.assert_np_array_equal(jac.row(1), data[1, :])

        jac = jac.transpose()
        self.assertEqual(jac.rows(), 3)

    def test_solve(self):
        jac = Jacobian.Random("jac", 7, "ee", "robot")
        twist = CartesianTwist.Random("ee", "robot")
        jac.solve(twist)

        data = np.random.rand(6, 7)
        jac.solve(data)

    def test_operators(self):
        jac = Jacobian.Random("test", 3, "ee", "robot")
        elem = jac[1, 2]
        jac[1, 2] = 0.1

        matrix = np.random.rand(3, 6)
        result = jac * matrix
        result = matrix * jac

        joint_velocities = JointVelocities.Random("test", 3)
        twist = jac * joint_velocities
        self.assert_np_array_almost_equal(joint_velocities.get_velocities(),
                                          (jac.pseudoinverse() * twist).get_velocities())

        wrench = CartesianWrench.Random("ee", "robot")
        torques = jac.transpose() * wrench

        pose = CartesianPose.Random("robot", "world")
        jac_in_world = pose * jac


if __name__ == '__main__':
    unittest.main()

File: source/state_representation/test/tests/space/test_jacobian.cpp
#include "state_representation/space/Jacobian.hpp"
#include "state_representation/exceptions/IncompatibleStatesException.hpp"
#include <gtest/gtest.h>

using namespace state_representation;
using namespace state_representation::exceptions;

TEST(JacobianTest, TestCreate) {
  Jacobian jac("robot", 7, "test");
  EXPECT_EQ(jac.get_type(), StateType::JACOBIAN);
  EXPECT_EQ(jac.rows(), 6);
  EXPECT_EQ(jac.cols(), 7);
  EXPECT_TRUE(jac.is_empty());
  EXPECT_EQ(jac.get_frame(), "test");
  EXPECT_EQ(jac.get_reference_frame(), "world");
  for (std::size_t i = 0; i < jac.cols(); ++i) {
    EXPECT_EQ(jac.get_joint_names().at(i), ("joint" + std::to_string(i)));
    EXPECT_EQ(jac.col(i).norm(), 0);
  }
  EXPECT_THROW(jac.set_joint_names(5), exceptions::IncompatibleSizeException);
  EXPECT_THROW(jac.set_joint_names(std::vector<std::string>{"j0"}), exceptions::IncompatibleSizeException);
}

TEST(JacobianTest, TestCreateWithVectorOfJoints) {
  Jacobian jac("robot", std::vector<std::string>{"j1", "j2"}, "test", "test_ref");
  EXPECT_EQ(jac.get_type(), StateType::JACOBIAN);
  EXPECT_EQ(jac.get_joint_names().at(0), "j1");
  EXPECT_EQ(jac.get_joint_names().at(1), "j2");
  EXPECT_EQ(jac.get_reference_frame(), "test_ref");
}

TEST(JacobianTest, TestSetData) {
  Jacobian jac("robot", 3, "test");
  jac.set_data(Eigen::MatrixXd::Random(6, 3));
  EXPECT_FALSE(jac.is_empty());
  for (std::size_t i = 0; i < jac.cols(); ++i) {
    EXPECT_GT(jac.col(i).norm(), 0);
  }
  bool except_thrown = false;
  try {
    jac.set_data(Eigen::MatrixXd::Random(7, 6));
  } catch (const IncompatibleSizeException& e) {
    except_thrown = true;
  }
  EXPECT_TRUE(except_thrown);
}

TEST(JacobianTest, TestRandomCreate) {
  Jacobian jac = Jacobian::Random("robot", 7, "test");
  EXPECT_EQ(jac.get_type(), StateType::JACOBIAN);
  EXPECT_FALSE(jac.is_empty());
  for (std::size_t i = 0; i < jac.cols(); ++i) {
    EXPECT_GT(jac.col(i).norm(), 0);
  }
}

TEST(JacobianTest, TestTranspose) {
  Jacobian jac = Jacobian::Random("robot", 7, "test");
  Jacobian jact = jac.transpose();
  EXPECT_EQ(jact.get_type(), StateType::JACOBIAN);

  EXPECT_EQ(jact.rows(), 7);
  EXPECT_EQ(jact.cols(), 6);

  for (std::size_t i = 0; i < jac.cols(); ++i) {
    EXPECT_TRUE(jac.col(i).isApprox(jact.row(i)));
  }
}

TEST(JacobianTest, TestMutltiplyWithEigen) {
  Jacobian jac = Jacobian::Random("robot", 7, "test");
  Eigen::MatrixXd mat1 = Eigen::MatrixXd::Random(7, 2);
  Eigen::MatrixXd res1 = jac * mat1;
  Eigen::MatrixXd res_truth = jac.data() * mat1;

  EXPECT_TRUE(res1.isApprox(res_truth));

  Eigen::MatrixXd mat2 = Eigen::VectorXd::Random(6, 1);
  bool except_thrown = false;
  try {
    Eigen::MatrixXd res2 = jac * mat2;
  } catch (const IncompatibleSizeException& e) {
    except_thrown = true;
  }
  EXPECT_TRUE(except_thrown);
}

TEST(JacobianTest, TestMutltiplyJacobian) {
  Jacobian jac = Jacobian::Random("robot", 7, "test");
  Jacobian jac2 = Jacobian::Random("robot", 7, "test");
  Eigen::MatrixXd res = jac * jac2.transpose();
  EXPECT_TRUE(res.isApprox(jac.data() * jac2.data().transpose()));

  // check with incorrect dimensions
  bool except_thrown = false;
  try {
    Eigen::MatrixXd res2 = jac * jac2;
  } catch (const IncompatibleSizeException& e) {
    except_thrown = true;
  }
  EXPECT_TRUE(except_thrown);

  // check with incorrect number of joints
  Jacobian jac3 = Jacobian::Random("robot", 6, "test");
  except_thrown = false;
  try {
    Eigen::MatrixXd res2 = jac * jac3.transpose();
  } catch (const IncompatibleStatesException& e) {
    except_thrown = true;
  }
  EXPECT_TRUE(except_thrown);

  // check with incorrect reference frames
  Jacobian jac4 = Jacobian::Random("robot", 7, "test", "frameA");
  except_thrown = false;
  try {
    Eigen::MatrixXd res2 = jac * jac4.transpose();
  } catch (const IncompatibleStatesException& e) {
    except_thrown = true;
  }
  EXPECT_TRUE(except_thrown);
}

TEST(JacobianTest, TestSolve) {
  Jacobian jac = Jacobian::Random("robot", 7, "test");
  Eigen::MatrixXd mat1 = Eigen::VectorXd::Random(7, 1);
  bool except_thrown = false;
  try {
    Eigen::MatrixXd res1 = jac.solve(mat1);
  } catch (const IncompatibleSizeException& e) {
    except_thrown = true;
  }
  EXPECT_TRUE(except_thrown);

  Eigen::MatrixXd mat2 = Eigen::VectorXd::Random(6, 1);
  Eigen::MatrixXd res2 = jac.solve(mat2);

  EXPECT_EQ(res2.rows(), 7);
  EXPECT_EQ(res2.cols(), 1);
}

TEST(JacobianTest, TestJointToCartesian) {
  Jacobian jac = Jacobian::Random("robot", 7, "test", "test_ref");
  JointVelocities jvel = JointVelocities::Random("robot", 7);
  CartesianTwist cvel = jac * jvel;

  EXPECT_EQ(cvel.get_name(), jac.get_frame());
  EXPECT_EQ(cvel.get_reference_frame(), jac.get_reference_frame());
  EXPECT_TRUE(cvel.data().isApprox(jac.data() * jvel.data()));
}

TEST(JacobianTest, TestCartesianToJoint) {
  Jacobian jac = Jacobian::Random("robot", 7, "test", "test_ref");
  CartesianTwist cvel = CartesianTwist::Random("test");

  bool except_thrown1 = false;
  try {
    JointVelocities jvel = jac.solve(cvel);
  } catch (const IncompatibleStatesException& e) {
    except_thrown1 = true;
  }
  EXPECT_TRUE(except_thrown1);
  cvel.set_reference_frame("test_ref");

  state_representation::JointVelocities jvel2;
  bool except_thrown2 = false;
  try {
    jvel2 = jac.pseudoinverse() * cvel;
  } catch (const IncompatibleSizeException& e) {
    except_thrown2 = true;
  }
  EXPECT_FALSE(except_thrown2);
  EXPECT_GT(jvel2.data().norm(), 0);
}

TEST(JacobianTest, TestChangeReferenceFrame) {
  Jacobian jac_in_test_ref = Jacobian::Random("robot", 7, "test", "test_ref");
  CartesianPose wTtest_ref = CartesianPose::Random("test_ref", "world");
  Jacobian jac_in_world = wTtest_ref * jac_in_test_ref;
  EXPECT_EQ(jac_in_world.get_reference_frame(), wTtest_ref.get_reference_frame());
  // use a proxy operation with a twist to check correctness
  CartesianTwist vel_in_world = CartesianTwist::Random("test", "world");
  CartesianTwist vel_in_test_ref = wTtest_ref.inverse() * vel_in_world;
  JointVelocities jt1 = jac_in_world.solve(vel_in_world);
  JointVelocities jt2 = jac_in_test_ref.solve(vel_in_test_ref);
  EXPECT_TRUE(jt1.data().isApprox(jt2.data()));
}
